### ISSUE TITLE
Minor enhancement request

### ISSUE THREAD

**Issue by brianavid** [Alxandr/SpotiFire#58]
Can I request a new one-line method as an enhancement please in Session.cpp. I did this myself several months ago, but forgot to push it back. So when I updated through nuget, it has disappeared (not a surprise!).

The method is simply:

    Link ^Session::GetLink(String ^link) {
        return Link::Create(this, link);
    }

This means that if I have an externally provided Spotify link (e.g. from the Web API), I can find the track or album in order to play it. Is anyone (Aleksander, Chris) in a position to do this and push the build through to nuget??

Thanks

Brian

**Comment by brianavid** [comment]
And of course in the Session.h file:

    virtual Link ^GetLink(String ^link) sealed;

**Comment by Alxandr** [comment]
I have no idea what it takes to get this project up and running anymore, as to what I need to install on my PC etc. If you want a new release pushed to nuget, send me a pull request with the changes. Include in the pull-request a version bump (IIRC, it's done by running the release script), and send me the resulting nuget package, and I can upload it for you.

**Comment by brianavid** [comment]
Thanks Aleksander,

I may do that later on. Though it sounds as though the project is winding down without future development, so it may not be worth it.

Meanwhile, I can (and will) use reflection in my calling code to cheat:

    static Link GetLink(
        this Session session,
        String id)
    {
        // Get Link.Create(Session,String)
        MethodInfo mInfo = typeof(Link).GetMethod(
            "Create",
            BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance,
            null,
            CallingConventions.Any,
            new Type[] { typeof(Session), typeof(String) },
            null);
        return mInfo?.Invoke(null, new object[] { session, id }) as Link;
    }

It's not the best solution - GetLink() should be in the API. But it will do for now.

**Comment by Alxandr** [comment]
More like it's been wound down a long time ago. I plan to redo it in C# eventually. Though I have no idea when that eventuality will come.

**Comment by brianavid** [comment]
Moving off-topic, and looking to the future, any re-write probably can't use libspotify exclusively as SpotiFire does now. That API is no longer in active development and does not support more recent capabilities such as "My Music". All new capabilities are exposed through the Web API only. 

It is for that reason that my own Spotify project (see https://www.youtube.com/watch?v=PSX-_iy29Pk)  uses a mixture of web API for browsing and SpotiFire for play queue management and song playing through NAudio. That is why I need GetLink() - to get an externally provided album or track link into SpotiFire to queue it for playing.

But using the Web API introduces significant issues of authentication and traffic throttling. So it will not be easy!

**Comment by eloekset** [comment]
According to the Libspotify documentation, there will come a new library for "other platforms", which means other than iOS and Android.
<img width="662" alt="2015-11-08_13-30-27" src="https://cloud.githubusercontent.com/assets/791721/11020342/17138368-861d-11e5-8c52-57e87e906012.png">

**Comment by brianavid** [comment]
The reflection idea was a non-starter - the non-public methods are not in the DLL :-(

Sent a pull-request with the new method added. Meanwhile I can use my private build

**Comment by Alxandr** [comment]
@brianavid Non-public methods are in the DLL, how else could they be used from within the code???

**Comment by brianavid** [comment]
That was my first thought. But in Link.h, Create() methods are in the `internal:` block and not emitted into the IL, so can't be found by reflection. But they are linked with Session.cpp in the DLL.

**Comment by Alxandr** [comment]
@brianavid They may be emitted as native methods, which can't be found by IL. Still strange though. Also, the new library which will come "later this year" (ie, late 2019) is probably a good point to rewrite SpotiFire again. The next one I plan to have xplat friendly (probably also CoreCLR).

**Comment by brianavid** [comment]
@Alxandr The use of reflection was not really a good idea anyway - I only tried it to save you the trouble of accepting the PR and uploading the nuget package. What we did eventually is by far the best outcome.

I am not holding my breath for any new APIs from Spotify. While libspotify is deprecated, it does seem to continue to work enough to serve track samples to an authenticated session within a custom player, which is all I need. That is a capability missing from the web API. The web API is great for browsing the catalog - and particularly "My Music". But the web API's authentication model is non-trivial to implement and I need to use a public URL on my web hosting to complete the handshake using my private key. I imagine that authentication in later APIs will be at least as tricky!
